drivers/hmis/app/graphql/mutations/publish_form_definition.rb

```diff
@@ -29,7 +29,7 @@ def resolve(id:)
       validation_errors = definition.validate_json_form
       return { errors: validation_errors } if validation_errors.any?
 
-      validation_errors = definition.validate_json_form
+      validation_errors = definition.validate_json_form(is_publishing: true)
```

Comment:
Just to make sure I have my head correctly wrapped around this:

The validator now will expect CDEDs to already exist for any form item that has a mapping with a `custom_field_key`. That could've thrown an error in the past, back when we used to add `custom_field_key` to the form and then generate the CDED (at first by hand, later using the rake task, but either way it was a two-step operation).

Now, though, as long as all forms currently published have their CDEDs generated already*, we should be OK, because on publish, the form builder will 
1) validate that all existing items that have mapping fields correspond to CDEDs that match, and then 
2) create new CDEDs for any new fields that need them, and add that mapping field to the json definition.

In other words, from now on, an item should never exist in an in-between state where it has a `custom_field_key` mapping, but no CDED.

Is there anything I'm missing?
*is this an okay assumption?

Reply:
Yeah that's correct. There is only one situation when we we expect the CDEDs don't exist, and that's when the form is in `draft` status.

IMO we should move the validation step below the CDED creation step, though, which means moving the call to `validate_json_form` down into the transaction where we have a `TODO`. That way we don't get into a situation where the `add_missing_custom_field_keys` code generates faulty CDEDs in a way that actually fails validation. (Currently we would never see that, since we validate before creating the new CDEDs).

I also think we should probably invert the boolean, i.e. make it so the  _default_ behavior when running the DefinitionValidator should be to validate everything, but include a toggle to skip certain validation.  That way we will still run the full validation [on seed](https://github.com/greenriver/hmis-warehouse/blob/26ef8c27b51dba80b065dfa9a3034722f165a15e/drivers/hmis/lib/hmis_util/json_forms.rb#L264) for example, which we want.

One option is to push this logic to the `definition.validate_json_form` logic itself:

```
  def validate_json_form
    # Skip validation of CustomDataElementDefinitions on draft form, because new CDEDs won't be created yet
    Hmis::Form::DefinitionValidator.perform(definition, role, skip_cded_validation: draft?)
  end
```
or, we could keep it exposed for the caller to specify:
```
  def validate_json_form(skip_cded_validation: false)
    Hmis::Form::DefinitionValidator.perform(definition, role, skip_cded_validation: skip_cded_validation)
  end

...

// in UpdateFormDefinition
# Skip validation of CustomDataElementDefinitions because this is a draft form, so new CDEDs won't be created yet
validation_errors = definition.validate_json_form(skip_cded_validation: true)
```

Reply:
All this makes sense, but I'm having a bunch of errors locally when I introduce the validation on seed. Probably my test DB is just out of date, so I wonder if this could be easily fixed by just running the rake task to generate cdeds? But the rake task to generate cdeds also has errors for me locally now. Maybe we could pair on this.